text/0000-hashing-builtin.md

```diff
@@ -0,0 +1,177 @@
+would look as nice and clean as this:
+
+```pony
+  fun hash(): USize =>
+    Hashing.hash_2(_field1, _field2)
+```
+
+# Detailed design
+
+## Move Hashable to Builtin
+
+We move the `Hashable` and `Hashable64` interfaces to the `builtin` package.
+
+## Make hashing objects convenient
+
+A primitive called `Hashing` will be introduced into the `builtin` package, alongside with a pure pony `Siphash` implementation.
+Currently libponyrt contains code for hashing byte-sequences with a C-implementation of Siphash
+in https://github.com/ponylang/ponyc/blob/master/src/libponyrt/ds/fun.c#L92 . As the implementation is already iterating over
+the array in 8-byte (4-byte for 32-bit platforms) increments, we can also use it in a streaming fashion, feeding it consecutive U64.
+This way we can feed it hashes of object fields and compute the sip-hash-2-4 of all the hashes of the object fields.
+
+When hashing objects based on their fields, we must take all those fields into account. They might be primitive types, objects with fields themselves,
+their hash might be using the fields identiy using the `digestof` operator etc. etc. So generating a hash boils down to a clever way to combine the hashes of all fields
+that should be considered. The hashes of the fields are chosen as a common denominator.
+And, ideally, those hashes should reflect all of the fields contents. 
+So if we build a hash of those hashes using a good enough hash function, we should, by transitivity, also get a good enough hash of the whole object in return.
+
+As far as I know, [SipHash-2-4](https://131002.net/siphash/) can be considered a good enough hash-function.
+It is also already in use for byte sequences and strings in Pony.
+
+This `Hashing` primitive will contain methods for hashing `ByteSeqs`, as e.g. `Array[U8]` does not implement `Hashable` but can be easily hashed,
+in fact all Arrays of primitive numeric types can and should be very easily hashed.
+It will contain methods for hashing `ReadSeq[Hashable #read]` by combining the element hashes into one hash value 
+using the [SipHash-2-4](https://131002.net/siphash/) implemention described above.
+It will also contain methods for conveniently combining hashes from different objects, like:
+
+```pony
+
+primitive Hashing
+  fun hash_3(a: Hashable box, b: Hashable box, c: Hashable box): USize =>
+    """
+    convenience function for combining the hashes of three object implementing `Hashable`
+    into one hash value.
+    """
+    ifdef ilp32 then
+      let sip = HalfSipHash24Streaming.create()
+      sip.update(a.hash().u32())
+      sip.update(b.hash().u32())
+      sip.update(c.hash().u32())
+      sip.finish()
+    else
+      let sip = SipHash24Streaming.create()
+      sip.update(a.hash().u64())
+      sip.update(b.hash().u64())
+      sip.update(c.hash().u64())
+      sip.finish()
+    end
+```
+
+It will contain such functions for 2 arguments up to 12 arguments, maybe more.
```

Comment:
One alternative to this approach that I want to mention is something like what Crystal does.

Every object in Crystal has a `hash` function that takes one argument of type `Hasher`. By default, that argument is a new instance of `Hasher`, which is a class similar in concept to what you've proposed here with the `Hashing` primitive. The difference is that it is implemented not as a primitive, but as an immutable/persistent data structure (similar to those `collections/persistent` in Pony).

If we took a similar approach, someone could implement hashing with an example looking something like this:

```pony
class MyClass
  let foo: Foo
  let bar: Bar
  let baz: Baz

  fun hash(): U32 =>
    hash_with(Hasher.create()).result()

  fun hash_with(hasher': Hasher): Hasher =>
    var hasher = hasher'
    hasher = this.foo.hash(hasher)
    hasher = this.bar.hash(hasher)
    hasher = this.baz.hash(hasher)
    hasher
```

The advantage to this kind of approach is that it won't require us to define N different `hash_1`, `hash_2`, `hash_3`, [...] functions as proposed here.

This approach also adds the ability to hash a dynamic number of inner fields to hash, rather than being limited to a static function signature with N arguments. For example, an implementation of a `List` style of data structure might want to hash everything present in the list, and in this approach it could be as simple as this:

```pony
  fun hash_with(hasher': Hasher): Hasher =>
    var hasher = hasher'
    for item in values() do
      hasher = this.foo.hash(hasher)
    end
    hasher
```

Reply:
I see the beauty in this approach, but it also feels like a weird indirection for the user.

I would not see a requirement for keeping the Hasher immutable/persistent, but would rather make it mutable, as i did with `Siphash24Streaming` here: https://github.com/ponylang/ponyc/compare/hashing-builtin?expand=1#diff-3fcbdccf4a6378f88bdb211e3838307bR9

```
fun hash_with(hasher: Hasher ref): Hasher ref =>
  this.foo.hash_with(hasher)
  this.bar.hash_with(hasher)
  this.baz.hash_with(hasher)
  hasher
``` 

Whatever is faster here.

So, this approach would change the interface `Hashable` to also include the `hash_with` method and would require to add a `Hasher` interface and implementation.

Also notice, that my `Hashing` implementation could conveniently solve the case of dynamic number of fields by adding a `hash_iter(iter: Iterator[Hashable box])` method.

Reply:
In general i think both approaches are interesting and would be willing to rewrite the RFC to incorporate @jemc's suggestion  as the RFCs main suggestion, if the people tend towards it.

Reply:
One note on performance - I'm not sure how the other differences would affect perforamance, but one difference that is worth noting is that the `Hashing.hash_2(x, y)` approach would need virtual calls, due to the calls being done on the `Hashable` interface arguments without a concrete type. The workaround for this would be to use type parameters to specialize for different types, but that leads to a much more verbose call site (i.e. `Hashing.hash_2[TypeX, TypeY](x, y)`).

I believe that the `Hasher` class approach wouldn't need any virtual calls, but maybe I'm missing something.

Reply:
Regarding mutable vs immutable `Hasher` class, I am not strongly opinionated, and would in general lean toward whichever approach is more performant. Probably the mutable approach is more performant in current Pony, since immutable objects need to allocated many more times, and in the absence of Heap-to-stack optimization, that's kind of a big deal.

Crystal addresses this by using a `struct` instead of a `class` for this type. In Crystal, a `struct` is passed "by value" instead of "by reference", so it's something like a Pony tuple, but with callable methods for it. This also explains why it is implemented as an immutable data structure, since that's the only sane way to deal with a "by value" object without major headaches.

Reply:
Now that I mentioned If we wanted to do absolutely zero-allocation, we could use a tuple ourselves. Something like:

```pony
type HashState is (U32, U32) // or whatever other kind of state we need...
primitive Hashing
  [...]
```

```pony
class MyClass
  let foo: Foo
  let bar: Bar
  let baz: Baz

  fun hash(): U32 =>
    Hashing.finish(hash_with(Hashing.begin())

  fun hash_with(state': HashState): HashState =>
    var state = state'
    state = this.foo.hash(state)
    state = this.bar.hash(state)
    state = this.baz.hash(state)
    state
```

This would probably be the most performant solution (if I had to guess), while also being generally nice in terms of developer experience.

Reply:
I like where this is going, @jemc !

Using SipHash-2-4 needs a state of `(U64, U64, U64. U64)` in 64 bit systems, and `(U32, U32, U32, U32)` on 32 bit systems. So we need some dispatching, or we work with USize all the way down. Anyway, it seems very much doable.